README.md
# StardewValleyCrabPot

This program helps you to find out how much gold per day you can make using crab pots in stardew valley.

# Installation
## Prerequrements
You need the program ``stack`` to be installed on your machine. See https://docs.haskellstack.org/en/stable/
for more details.

## Building and run

1. Download or clone this repository onto your machine. If you have git you can use

   ``> git clone https://github.com/guslav/StardewValleyCrabPot.git``
   
2. Switch into the new folder

   ``> cd StardewValleyCrabPot``
3. Compile and run

   ``> stack run``


CHANGELOG.md
# Changelog for `StardewValleyCrabPot`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/StardewValley.hs
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE DeriveFoldable #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module StardewValley
    ( Place (..)
    , Trash (..)
    , Ocean (..)
    , Fresh (..)
    , Item (..)
    , Farming (..)
    , Fishing (..)
    , ProbTree (..)
    , MyPercentage (..)
    , MyRational (..)
    , everything
 --   , showPercentage
 --   , showRational
    , showTree
    , showTreeProbs
    , trashTree
    , recyclingGoods
    , multiplesOfRecyclingGoods
    , recycleTrashTree
    , recyclingGoodsTree
    , oceanTree
    , freshTree
    , probabilityIntoLeafs
    , probabilityOfItem
    , eventProbability
    , average
    , getProbabilityTree
    , isFish
    , profit
    , profitsIntoLeafs
    , averageProfit
    , trimShow
    , table
    , tableRecyclings
    , roundFraction
    ) where

import Data.List (intercalate, findIndices)
import Data.Ratio (numerator, denominator)
import Control.Monad (ap)
import Data.Bifunctor (second)

data Place = InOcean | InFresh 
  deriving (Show, Eq, Enum, Bounded)

data Trash = Junk | Driftwood | Newspaper | CD | Glasses 
  deriving (Show, Eq, Enum, Bounded)
data Ocean = Lobster | Clam | Crab | Cockle | Mussel | Shrimp | Oyster 
  deriving (Show, Eq, Enum, Bounded)
data Fresh = Crayfish | Snail | Periwinkle
  deriving (Show, Eq, Enum, Bounded)
data RecyclingGood = Stone | Coal | IronOre | Wood | Torch | Cloth | Quarz
  deriving (Show, Eq, Enum, Bounded)

data Item = Trash (Int, RecyclingGood) | Ocean Ocean | Fresh Fresh
  deriving (Eq)

data Fishing = NoFishing | Fisher | Angler | Mariner | MarinerAngler  
  deriving (Show, Eq, Enum, Bounded)

data Farming = NoRecycling | NoFarming | Rancher | Artisan
  deriving (Show, Eq, Enum, Bounded)

instance Show Item where
  show (Trash t) = show t
  show (Ocean t) = show t
  show (Fresh t) = show t

data ProbTree d a = Node [(d, ProbTree d a)] | Leaf a
  deriving (Show, Foldable, Functor)

instance Applicative (ProbTree d) where
  pure  = Leaf
  (<*>) = ap

instance Monad (ProbTree d) where
  Leaf x >>= f = f x
  Node l >>= f = Node $ fmap (second (>>= f) ) l

-- | 'everything' is a little hack, to get a list of every constructor for a given type.
everything :: (Enum a, Bounded a) => [a]
everything = [minBound .. maxBound]

showTree :: forall a d. (Show a, Show d) => ProbTree d a -> String
showTree = intercalate "\n" . map fst . showTree'
  where
    showTree' :: ProbTree d a -> [(String, Bool)]
    showTree' (Leaf x) = [(show x, False)]
    showTree' (Node l) = do
        (p,t) <- l
        let subLines = showTree' t
        let lastArrowLine = case findIndices snd subLines of
              [] -> 0
              l'  -> last l'
        let indent = zipWith (f lastArrowLine) [0..] subLines
        (show p, True) : indent
    f _ _ (line, True) = ("+-->" ++ line, False)
    f lastArrowLine i (line, False)
      | i < lastArrowLine = ("|   " ++ line, False)
      | otherwise         = ("    " ++ line, False)

newtype LeafLine d a = LeafLine (d, a)

instance (Show a, Show d, RealFrac d) => Show (LeafLine d a) where
  show (LeafLine (p, x)) = show x ++ " => " ++ show p

newtype MyRational = MyRational Rational
  deriving (Num, Eq, Fractional, RealFrac, Real, Ord)

newtype MyPercentage = MyPercentage Rational
  deriving (Num, Eq, Fractional, RealFrac, Real, Ord)

instance Show (MyRational) where
  show (MyRational d) = "(" ++ show (numerator d) ++ "/" ++ show (denominator d) ++ ") \8776 " ++ roundFraction 2 d 

instance Show (MyPercentage) where
  show (MyPercentage d) = "(" ++ show (numerator d) ++ "/" ++ show (denominator d) ++ ") \8776 " ++ roundFraction 1 (100*d) ++ "%"

showTreeProbs :: (Show a, Show d, RealFrac d) => ProbTree d a -> String
showTreeProbs t = showTree $ LeafLine <$> probabilityIntoLeafs t

evenDistributed :: Fractional d => [a] -> ProbTree d a
evenDistributed l = Node $ map ((probability, ) . Leaf ) l
  where
    probability = intRecip $ length l

intRecip :: Fractional d => Int -> d
intRecip = recip . fromInteger . toEnum

trashTree :: Fractional d => ProbTree d Trash
trashTree = evenDistributed everything

recyclingGoods :: Fractional d => Trash -> [(d, RecyclingGood)]
recyclingGoods Junk      = [(49/100, Stone),(30/100, Coal), (21/100, IronOre)]
recyclingGoods Driftwood = [(75/100, Wood),(25/100, Coal)]
recyclingGoods Newspaper = [(90/100, Torch),(10/100, Cloth)]
recyclingGoods CD        = [(1, Quarz)]
recyclingGoods Glasses   = [(1, Quarz)]

multiplesOfRecyclingGoods :: RecyclingGood -> [Int]
multiplesOfRecyclingGoods Stone   = [1,2,3] 
multiplesOfRecyclingGoods Coal    = [1,2,3]
multiplesOfRecyclingGoods IronOre = [1,2,3]
multiplesOfRecyclingGoods Wood    = [1,2,3]
multiplesOfRecyclingGoods Torch   = [3]
multiplesOfRecyclingGoods Cloth   = [1]
multiplesOfRecyclingGoods Quarz   = [1]

recycleTrashTree :: Fractional d => Trash -> ProbTree d (Int, RecyclingGood)
recycleTrashTree trash = do
    recyclingGood <- Node $ second Leaf <$> recyclingGoods trash
    let multiples = multiplesOfRecyclingGoods recyclingGood
    leaf <- Node $ (intRecip (length multiples),) <$> Leaf <$> (, recyclingGood) <$> multiples
    return leaf

recyclingGoodsTree :: Fractional d => ProbTree d (Int, RecyclingGood)
recyclingGoodsTree = trashTree >>= recycleTrashTree

splitTree :: Fractional d => d -> ProbTree d a -> ProbTree d a -> ProbTree d a
splitTree p t t' = Node [ (p, t), (1 - p, t') ]

chainTree :: Fractional d =>  ProbTree d a -> [(d,a)] -> ProbTree d a
chainTree = foldr (\(d, a) t -> splitTree d (Leaf a) t )

noMarinerTree :: Fractional d => (a -> Item) -> [(d,a)] -> ProbTree d Item
noMarinerTree f catchProbs = splitTree (80 / 100) catch trashT
  where
    trashT = fmap Trash recyclingGoodsTree
    catch = chainTree trashT $ second f <$> catchProbs


{- Fish.xnb
  "715": "Lobster/trap/.05/688 .45 689 .35 690 .35/ocean/2/20",
  "717": "Crab/trap/.1/684 .45/ocean/1/20",
  "723": "Oyster/trap/.15/682 .35/ocean/1/5",
  "372": "Clam/trap/.15/681 .35/ocean/1/5",
  "720": "Shrimp/trap/.2/681 .35/ocean/1/3",
  "718": "Cockle/trap/.3/680 .2/ocean/1/5",
  "719": "Mussel/trap/.35/683 .15/ocean/1/5",
  "721": "Snail/trap/.25/680 .35/freshwater/1/5",
  "716": "Crayfish/trap/.35/682 .4/freshwater/1/8",
  "722": "Periwinkle/trap/.55/681 .35/freshwater/1/3",
-}

oceanTree :: Fractional d => ProbTree d Item
oceanTree = noMarinerTree Ocean [ ( 5 / 100, Lobster)
                                , (10 / 100, Crab)
                                , (15 / 100, Oyster)
                                , (15 / 100, Clam)
                                , (20 / 100, Shrimp)
                                , (30 / 100, Cockle)
                                , (35 / 100, Mussel)
                                ]


freshTree :: Fractional d => ProbTree d Item
freshTree = noMarinerTree Fresh [ (25 / 100, Snail)
                                , (35 / 100, Crayfish)
                                , (55 / 100, Periwinkle)
                                ]

probabilityIntoLeafs :: Fractional d => ProbTree d a -> ProbTree d (d, a)
probabilityIntoLeafs = probabilityIntoLeafs' 1
  where
    probabilityIntoLeafs' :: Fractional d => d -> ProbTree d a -> ProbTree d (d, a)
    probabilityIntoLeafs' p (Leaf a) = Leaf (p, a)
    probabilityIntoLeafs' p (Node l) = Node $ map (\(d, t') -> (d, probabilityIntoLeafs' (p * d) t')) l

probabilityOfItem :: (Eq a, Fractional d) => ProbTree d a -> a -> d
probabilityOfItem tree x = eventProbability tree (==x)

eventProbability :: Fractional d => ProbTree d a -> (a -> Bool) -> d
eventProbability tree f = sum $ insertProbs <$> probabilityIntoLeafs tree
  where
    insertProbs (p, x)
      | f x = p
      | otherwise = 0

average :: Fractional d => ProbTree d d -> d
average t = sum $ uncurry (*) <$> probabilityIntoLeafs t

getProbabilityTree :: Fractional d => Place -> Fishing -> ProbTree d Item
getProbabilityTree place skill = case (isMariner, place) of
        (False, InOcean) -> oceanTree
        (False, InFresh) -> freshTree
        (True, InOcean)  -> fmap Ocean $ evenDistributed $ (everything :: [Ocean])
        (True, InFresh)  -> fmap Fresh $ evenDistributed $ (everything :: [Fresh])
  where
    isMariner = elem skill [Mariner, MarinerAngler]

-- | This Function checks if a certain Item can be converted into sashimi.
isFish :: Item -> Bool
isFish (Trash _) = False
isFish (Ocean Clam) = False
isFish _ = True

recyclingGoodValue :: Integral i => Farming -> RecyclingGood -> i
recyclingGoodValue NoRecycling _ = 0
recyclingGoodValue _ Stone   = 2
recyclingGoodValue _ Coal    = 15
recyclingGoodValue _ IronOre = 10
recyclingGoodValue _ Wood    = 2
recyclingGoodValue _ Torch   = 5 
recyclingGoodValue _ Quarz   = 50
recyclingGoodValue NoFarming Cloth = 470
recyclingGoodValue Rancher Cloth = 564
recyclingGoodValue Artisan Cloth = 658

baseValue :: Integral i => Item -> i 
baseValue ( Trash _ ) = 0
baseValue (Ocean o) = case o of
    Lobster -> 120
    Clam   -> 50
    Crab   -> 100
    Cockle -> 50
    Mussel -> 30
    Shrimp -> 60
    Oyster -> 40
baseValue (Fresh f) = case f of
    Crayfish   -> 75
    Snail      -> 65
    Periwinkle -> 20

profit :: Integral i => Fishing -> Farming -> Bool -> Item -> i
profit fi fa sashimi item = case item of
  Trash (n,x) -> toEnum n * recyclingGoodValue fa x
  _       -> intoSashimi $ applyFishingProfession $ baseValue item
  where
    intoSashimi :: Integral i => i -> i
    intoSashimi
        | sashimi && isFish item = max 75
        | otherwise              = id
    applyFishingProfession :: Integral i => i -> i
    applyFishingProfession = case (isFish item, fi) of
        (False, _)  -> id
        (_, Fisher) -> \i -> div (i * 5) 4
        (_, Angler) -> \i -> div (i * 3) 2
        (_, MarinerAngler) -> \i -> div (i * 3) 2
        _ -> id

profitsIntoLeafs :: Integral i => Fishing -> Farming -> Bool -> ProbTree d Item -> ProbTree d (Item, i)
profitsIntoLeafs fi fa sashimi = fmap (\item -> (item, profit fi fa sashimi item))

averageProfit :: (Fractional d) => Fishing -> Farming -> Bool -> ProbTree d Item -> d
averageProfit fi fa sashimi tree = average $ fromInteger . snd <$> profitsIntoLeafs fi fa sashimi tree

trimShow :: Show a => Int -> a -> String
trimShow n str = take n $ show str ++ repeat ' '

table :: String
table = intercalate "\n" $ do
  place <- everything
  sashimi <- everything
  fishing <- everything
  let farming = NoRecycling
  let tree = getProbabilityTree place fishing :: ProbTree Rational Item
  let avprofit = averageProfit fishing farming sashimi tree :: Rational
  return $ intercalate " " $ [trimShow 7 place, trimShow 10 fishing, trimShow 10 farming, trimShow 5 sashimi, roundFraction 2 avprofit, show avprofit]

tableRecyclings :: String
tableRecyclings = intercalate "\n" $ do
  place <- everything
  sashimi <- everything
  fishing <- everything
  let tree = getProbabilityTree place fishing :: ProbTree Rational Item
  let avprofit f = averageProfit fishing f sashimi tree :: Rational
  let profits = roundFraction 2 <$> subtract (avprofit NoRecycling) <$> avprofit <$> [NoRecycling, NoFarming, Rancher, Artisan]
  return $ intercalate " " $ [trimShow 7 place, trimShow 10 fishing, trimShow 5 sashimi] ++ profits

roundFraction :: forall d. RealFrac d => Int -> d -> String
roundFraction n f = show b ++ "." ++ fillLeft n '0' (show c)
  where
    (b,a)  = properFraction f :: (Integer, d)
    (c,_)  = properFraction (a * 10^n) :: (Integer, d)

fillLeft :: Int -> a -> [a] -> [a]
fillLeft n x l = replicate (n - length l) x ++ l


app/Main.hs
module Main (main) where

import Data.IORef
--import Data.Ratio ((%))
import Data.Char (toLower)
import Data.List (intercalate)

import System.IO (hSetBuffering, BufferMode(..) , stdout)

import Control.Monad.Writer.Lazy (tell, execWriter)

import System.Directory (createDirectoryIfMissing)

import StardewValley

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering
    place   <- newIORef InOcean
    farming <- newIORef NoRecycling
    fishing <- newIORef NoFishing
    sashimi <- newIORef False
    let printcase' = printcase place farming fishing sashimi
    putStrLn ""
    putStrLn ""
    putStrLn "Welcome to the stardew valley crab pot calculator. Type 'help' if you want to see a list of the available commands."
    let main' = do
            p <- readIORef place
            a <- readIORef farming
            i <- readIORef fishing
            s <- readIORef sashimi
            putStr "\nStardewValley> "
            command <- words <$> map toLower <$> getLine
            let tree = getProbabilityTree p i :: ProbTree Rational Item
            let avprofit = averageProfit i a s tree
            case command of
                "q":_      -> putStrLn "bye!"
                "quit":_   -> putStrLn "bye!"
                "exit":_   -> putStrLn "bye!"
                "help":_   -> help >> main'
                "printcase":_ -> printcase' >> main'
                "printtable":_ -> printtable >> main'
                "onlinetable":_      -> putStrLn onlineTable >> main'
                "onlinetabletrash":_ -> putStrLn onlineTableTrash >> main'
                "set":p1:l    -> do
                    case p1 of
                        "place" -> case l of
                            ["ocean"]      -> writeIORef place InOcean
                            ["freshwater"] -> writeIORef place InFresh
                            _ -> do
                                putStrLn "Wasn't able to read the place at which you want to fish with your crab pot. Available options are 'ocean' and 'freshwater'."
                                putStrLn "Nothing changed.\n"
                        "farming" -> case l of
                            ["norecycling"] -> writeIORef farming NoRecycling
                            ["no","recycling"] -> writeIORef farming NoRecycling
                            ["nofarming"] -> writeIORef farming NoFarming
                            ["no","farming"] -> writeIORef farming NoFarming
                            ["rancher"] -> writeIORef farming Rancher
                            ["artisan"] -> writeIORef farming Artisan
                            _ -> do
                                putStrLn "Wasn't able to read your farming skill. Available options are 'norecycling', 'nofarming', 'rancher' and 'artisan'."
                                putStrLn "Nothing changed.\n"
                        "fishing" -> case l of
                            ["nofishing"] -> writeIORef fishing NoFishing
                            ["no","fishing"] -> writeIORef fishing NoFishing
                            ["fisher"] -> writeIORef fishing Fisher 
                            ["angler"] -> writeIORef fishing Angler 
                            ["mariner"] -> writeIORef fishing Mariner 
                            ["marinerangler"] -> writeIORef fishing MarinerAngler 
                            ["mariner", "angler"] -> writeIORef fishing MarinerAngler 
                            ["mariner&angler"] -> writeIORef fishing MarinerAngler 
                            _ -> do
                                putStrLn "Wasn't able to read your fishing skill. Available options are 'nofishing', 'fisher', 'angler', 'mariner', 'marinerangler'"
                                putStrLn "Nothing changed.\n"
                        "sashimi" -> case l of
                            ["false"] -> writeIORef sashimi False
                            ["no"]    -> writeIORef sashimi False
                            ["true"]  -> writeIORef sashimi True
                            ["yes"]   -> writeIORef sashimi True
                            _ -> do
                                putStrLn "Wasn't able to change if you want to make every fish below 75g into sashimi before selling. Available options are 'true' and 'false'. (You can also write 'yes' or 'no'.)"
                                putStrLn "Nothing changed.\n"
                        _ -> do
                            putStrLn "Wasn't able to read what variable you are intending to change. Available options are 'place', 'farming', 'fishing' and 'sashimi'."
                            putStrLn "Nothing changed.\n"
                    printcase'
                    main'
                "profit":_ -> do
                    putStrLn $ "You would make a profit of (" ++ show avprofit ++ ") \8776 " ++ roundFraction 2 avprofit ++ " gold every day."
                    putStrLn $ "Note, that haskell is using '%' for fractions."
                    putStrLn $ "If you want a reasoning for that number type command 'proof'."
                    main'
                "proof":_  -> putStrLn (showProof p a i s) >> main'
                "writeproofs":_ -> writeProofs >> main'
                _ -> putStrLn "Unknown command. Type 'help' if you want a list of valid commands." >> main'
    main'        

writeProofs :: IO ()
writeProofs = do
    let folder = "calculations"
    createDirectoryIfMissing False folder
    allCases $ \place farming fishing sashimi -> do
            let filename = (++ ".txt") $ intercalate "_" $ [show place, show farming, show fishing, showSashimi sashimi]
            let content = filename ++ "\n" ++ showProof place farming fishing sashimi
            writeFile (folder ++ "/" ++ filename) content
  where
    allCases :: (Enum a, Bounded a, Enum b, Bounded b, Enum c, Bounded c, Enum d, Bounded d) => (a -> b -> c -> d -> IO ()) -> IO ()
    allCases f = sequence_ $ f <$> everything <*> everything <*> everything <*> everything
    showSashimi :: Bool -> String
    showSashimi s
      | s = "Sashimi"
      | otherwise = "NoSashimi"

showProof :: Place -> Farming -> Fishing -> Bool -> String
showProof place farming fishing sashimi = execWriter $ do
    let tellLn = tell . (++"\n")
    let tree = getProbabilityTree place fishing :: ProbTree MyPercentage Item
    let avprofit = fromRational $ toRational $ averageProfit fishing farming sashimi tree :: MyRational
    tellLn $ "\n\nYou would make a profit of " ++ show avprofit ++ " gold every day."
    tellLn $ "The number can be obtained by the following probability tree. Just add the product of probability and selling value of any leaf in the tree together and you will get the number " ++ roundFraction 2 avprofit ++ "."
    if farming == NoRecycling
      then tellLn "NOTE: You are not selling the trash. Therefore all the recycling items (Stone, Coal, Wood, IronOre, Cloth, Torch) get a selling value of 0." 
      else return ()
    tellLn ""
    tellLn $ showTree $ intoProofLines fishing farming sashimi tree

printtable :: IO ()
printtable = do
  putStrLn $ trimmer ["Place", "FishingSkill", "FarmingSkill", "Sashimi", "profit", "exact average profit (fraction)"]
  allCases summulateCertain
  where
    allCases :: (Enum a, Bounded a, Enum b, Bounded b, Enum c, Bounded c, Enum d, Bounded d) => (a -> b -> c -> d -> IO ()) -> IO ()
    allCases f = sequence_ $ f <$> everything <*> everything <*> everything <*> everything
    trimmer :: [String] -> String
    trimmer [a,b,c,d,e,f] = intercalate " " [trim 7 a, trim 12 b, trim 12 c, trim 8 d, trim 8 e, f]
    trimmer _ = error "something didnt work."
    trim :: Int -> String -> String
    trim n str = take n $ str ++  repeat ' '
    summulateCertain place fishing sashimi farming  = do
        let tree = getProbabilityTree place fishing :: ProbTree Rational Item
        let avprofit = averageProfit fishing farming sashimi tree :: Rational
        putStrLn $ trimmer [show place, show fishing, show farming, show sashimi, roundFraction 2 avprofit, show avprofit]

onlineTable :: String
onlineTable = (top ++) $ (++ "|}") $ do
    place <- everything
    sashimi <- everything
    left place sashimi ++ right place sashimi
  where
    top = "{| class=\"wikitable\"\n|\n! [[Luremaster]] or no Fishing Profession \n! [[Fisher]]\n! [[Fisher|Angler]]\n! [[Mariner]]\n! [[Fisher|Angler]] & [[Mariner]]\n"
    left :: Place -> Bool -> String
    left place sashimi = "|- \n! " ++ showPlace place ++ s ++ "\n"
        where s = if sashimi then " + [[Sashimi]]" else ""
    right :: Place -> Bool -> String
    right place sashimi = "| " ++ cell NoFishing ++ "\n| " ++ cell Fisher ++ "\n| " ++ cell Angler ++ "\n| " ++ cell Mariner ++ "\n| " ++ cell MarinerAngler ++ "\n"
      where
        cell :: Fishing -> String
        cell fishing = roundFraction 2 $ (averageProfit fishing NoRecycling sashimi $ getProbabilityTree place fishing :: Rational)

onlineTableTrash :: String
onlineTableTrash = (top ++) $ (++ "|}") $ do
    place <- everything
    left place ++ right place
  where
    top = "{| class=\"wikitable\"\n|\n! [[Agriculturist]] or no Farming Profession  \n! [[Rancher]]\n! [[Artisan]]\n"
    left :: Place -> String
    left place = "|- \n! " ++ showPlace place ++ "\n"
    right :: Place -> String
    right place = "| " ++ cell NoFarming ++ "\n| " ++ cell Rancher ++ "\n| " ++ cell Artisan ++ "\n"
      where
        cell :: Farming -> String
        cell farming =  roundFraction 2 $ prof farming - prof NoRecycling
        prof :: Farming -> Rational
        prof farming = (averageProfit NoFishing farming False $ getProbabilityTree place NoFishing :: Rational)

showPlace :: Place -> String
showPlace InOcean = "Ocean"
showPlace InFresh = "Freshwater"

printcase :: IORef Place -> IORef Farming -> IORef Fishing -> IORef Bool -> IO ()
printcase place farming fishing sashimi = do
    p <- readIORef place
    a <- readIORef farming
    i <- readIORef fishing
    s <- readIORef sashimi
    putStrLn "This is the actual state to look at:"
    case p of
        InOcean -> putStrLn "The crab pot is inside the ocean."
        InFresh -> putStrLn "The crab pot is inside the freshwater."
    case a of
        NoRecycling -> putStrLn "You do not plan to recycle the trash. Therefore your farming profession doesnt matter."
        NoFarming   -> putStrLn "You want to recycle your trash in a recycling machine. Your farming skill doesnt make the cloth more valuable. (e.g. it isnt rancher or artisan)"
        Rancher     -> putStrLn "You want to recycle your trash in a recycling machine and your farming skill is rancher. Therefore cloth is 20% more valueable."
        Artisan     -> putStrLn "You want to recycle your trash in a recycling machine and your farming skill is artisan. Therefore cloth is 40% more valueable."
    case i of
        NoFishing     -> putStrLn "You have the lurelaster profession or no fishing profession."
        Fisher        -> putStrLn "You have the fishing profession but not the angler profession."
        Angler        -> putStrLn "You have the angler profession."
        Mariner       -> putStrLn "You have the mariner profession."
        MarinerAngler -> putStrLn "You have the mariner profession but you will switch to the angler profession when you sell the fish."
    case s of
        False -> putStrLn "You will NOT turn the fish with value less then 75g into sashimi."
        True  -> putStrLn "You will turn the fish with value less then 75g into sashimi."

intoProofLines :: (Fractional d) => Fishing -> Farming -> Bool -> ProbTree d Item -> ProbTree d (ProofLine d)
intoProofLines fishing farming sashimi tree = (\(prob, item) -> ProofLine prob item (profit fishing farming sashimi item)  )  <$> probabilityIntoLeafs tree

data ProofLine d = ProofLine d Item Integer

instance (RealFrac d, Show d) => Show (ProofLine d) where
    show (ProofLine probability item value) = showItem item ++ " has a probability of " ++ show probability ++ " and a selling value of " ++ show value ++ ". Product of both: " ++ show prod
      where
        prod = fromRational $ toRational $ probability * fromInteger value :: MyRational
        showItem (Trash (n,name) ) = show n ++ "x" ++ show name
        showItem (Ocean o) = show o
        showItem (Fresh f) = show f



help :: IO ()
help = do
    putStrLn "This program helps you to find out how much gold per day you can make using crab pots in stardew valley."
    putStrLn "Initialy the program assumes:"
    putStrLn "    1.) The crab pot is placed in the ocean."
    putStrLn "    2.) That you dont sell the trash / recycling goods."
    putStrLn "    3.) That you dont have any fishing profession."
    putStrLn "    4.) That you dont turn the fish into sashimi before selling."
    putStrLn "    5.) That you do NOT want to include the price of fishing bait."
    putStrLn "You can change the presets of 1 - 4 by using the 'set' command."
    putStrLn ""
    putStrLn "The following commands are available:  'exit', 'q', 'quit', 'help', 'printcase', 'profit', 'proof', 'printtable', 'onlinetable', 'onlinetabletrash', 'set'"
    putStrLn "    exit: Ends this program."    
    putStrLn "    q:    Same as exit."    
    putStrLn "    quit: Same as exit."
    putStrLn "    help: Prints the text you are currently reading."
    putStrLn "    printcase: Shows the assumptions for the calculation. (See 1. - 5.) "
    putStrLn "    profit: Shows you how much profit you would make using a crab pot. (depends on the assumptions)"
    putStrLn "    proof: Gives a reasoning for the profit by printing the entire probability tree. "
    putStrLn "    printtable: prints a table with the profits under each possible assumption set (see 1. - 5. above) "
    putStrLn "    onlinetable: prints a String, that makes a table for stardewvalleywiki.com. See 'https://stardewvalleywiki.com/Talk:Crab_Pot#Calculating_the_gold_per_day' to see an example."
    putStrLn "    onlinetabletrash: prints a String, that makes a table for stardewvalleywiki.com. See 'https://stardewvalleywiki.com/Talk:Crab_Pot#Calculating_the_gold_per_day' to see an example."
    putStrLn "    set [variable] [value]: Changes one of the assumptions. Valid values for variable are 'place', 'farming', 'fishing' and 'sashimi'."
    putStrLn ""
    putStrLn "     > set place ocean           (preset)"
    putStrLn "     > set place freshwater"
    putStrLn ""
    putStrLn "    If you plan to sell the recycling goods, it is important what farming profession you got, because the selling price of cloth depends on that."
    putStrLn "    You can also pick 'nofarming' if you will sell the recycling goods but selling cloth for its standard price of 470g."
    putStrLn "     > set farming norecycling   (preset)"
    putStrLn "     > set farming nofarming"
    putStrLn "     > set farming rancher"
    putStrLn "     > set farming artisan"
    putStrLn ""
    putStrLn "    Select your fishing perfection. 'marinerangler' means, that you pick mariner to collect the fish but switching for one day to angler by the " 
    putStrLn "    dog statue to sell for a better price. The cost of 10.000g is excluded. collect the fish for a while in a box."
    putStrLn "     > set fishing nofishing     (preset)"
    putStrLn "     > set fishing fisher"
    putStrLn "     > set fishing angler"
    putStrLn "     > set fishing mariner"
    putStrLn "     > set fishing marinerangler"
    putStrLn ""
    putStrLn "    Select if you will turn all the fish with a selling price less then 75g into sashimi before selling. "
    putStrLn "     > set sashimi false         (preset)"
    putStrLn "     > set sashimi true"
